README.md
# x11mirror-client

## Description

This is an X11 program which retrieves an image (a screenshot)
of the specified window or a whole display, saves it to a file
and, possibly, sends it to a remote server.

A video is worth a thousand words:
[video demonstration on YouTube](https://youtu.be/4Kz16FlFcjQ) ~7min.

The [x11mirror-server][1] is a HTTP server which demonstrated
in the video above. It is a complementary part for `x11mirror-client`.


## Features

* Don't create any windows for a work.
* Supports zlib to compress huge XWD format.
* Supports curl to upload files.


## Dependecies

* X11 development headers (X11, Xcomposite, Xrender, Xdamage)
* zlib (optional)
* curl (optional)
* GNU make (to build)
* C99 compiler (gcc, clang, etc)


## Build

Currently tested only on GNU/Linux. By default the program is building 
with CURL & ZLIB support. Use `make help` to see general options:

```
% make help
WITH_CURL=YES|NO - enable/disable curl support, default: YES
WITH_ZLIB=YES|NO - enable/disable zlib support, default: YES
ENABLE_DELAY=YES|NO - use delay between screenshots, default: YES
ENABLE_DEBUG=YES|NO - be verbose and print debug information, default: NO
ENABLE_ERRORS=YES|NO - print errors to STDERR, default: YES
```

Additional options that should be passed via `CFLAGS`:

* `_DELAY_SEC`, `_DELAY_NSEC` - default is `1` and `0` respectively.
* `_UPLOAD_URL` - default is `"http://localhost:8888/"`.
* `_OUTPUT_FILE` - default is `/tmp/x11mirror.xwd`.
* `_ZLIB_DEFAULT_LEVEL` - default is 7.

Note that the default value is `CFLAGS=-O2`. Please, see [main.c](/main.c)
and [Makefile](/Makefile) for details.

An example how to change the uploading URL:

```
% make CFLAGS="-Os -D_UPLOAD_URL=\\\"http://127.0.0.1:8888\\\""
```


### Build dependencies

To build the application you have to install development packages on your
system. The list below shows packages to install on Void Linux:

* libX11-devel
* libXcomposite-devel
* libXrender-devel
* libXdamage-devel

Note that on your system the package names could differ from the list above.

Optionally you may install zlib and curl libriries too:
* zlib-devel
* libcurl-devel


## Usage

### Limitations

* The program must be run on the same machine where X11 xorg-server is
  running.
* X11 server must support next extentions: composite, RENDER, FIXES, DAMAGE.
  On modern GNU/Linux distributives they are all supported and available.


### General usage

Please see options below to understand basics:

```
% ./x11mirror-client --help
Usage: ./x11mirror-client [-w window] [OPTIONS]
Options:
  --help, -h        print this help
  -d display        connection string to X11
  -o output         output filename, default: /tmp/x11mirror.xwd
  -w window         target window id, default: root
  -S                enable X11 synchronization, default: disabled
  -u URL            an URL to send data
  -U                enable uploading, default: disabled
  -z                enable gzip (zlib) compression, default: disabled
  -Z                zlib compression level (1-9), default: 7
  -D                delay between making screenshots in milliseconds
```

Quickstart command:

```
% ./x11mirror-client
```

By default the latest screenshot file could been found in `/tmp/x11mirror.xwd`
(not compressed) and it will be updated every 1 second.

You can stop the program at any time by pressing `CTRL+C` on keyboard.


### Supported modes

The *x11mirror-client* can be built in different ways:

* Without CURL support
* Without ZLIB support

If the program was built with CURL support it can send
screenshots to a remote HTTP server in automatic mode.

Without CURL support the program just saves a screenshot to
the specified file.

Default values are:

* Remote server URL: `http://localhost:8888/`
* Output file: `/tmp/x11mirror.xwd`

In any case the program will store a screenshot to a file. The original
idea was to keep the screenshot into the program's memory.
May be I will implement it later.

Because of fact that the file will be frequently updated I advise you to
use `tmpfs`.


#### Why ZLIB?

The XWD format consumes a lot of disk space. To understand this better,
please look at these numbers:

```
-rw------- 1 xxx xxx 4.1M Mar 23 22:55 xmc.xwd
-rw------- 1 xxx xxx  74K Mar 23 22:41 xmc.xwd.gz
```

The files contain the same screenshot of the whole display with the
resolution 1366x768 pixels.


#### How do I convert XWD to PNG, JPEG?

Well, it is quite common question in Internet. You have to use `convert`
utility from ImageMagick package. For instance, to convert XWD to JPG
just type in console the next command:

```
% convert xwd:input_file.xwd jpg:output_file.jpg
```

The `convert` utility understands gzipped XWD files too, so you don't
need to `gunzip` them:

```
% convert xwd:input_file.xwd.gz jpg:output_file.jpg
```

One note to be mention: the ImageMagick on your system must be built
with X11 support. You may check if its ok by the next command:

```
% convert -list format | grep XWD
      XWD* XWD       rw-   X Windows system window dump (color)
```


#### Window ID

Each X11's window has its own identifier (window id). The `xwininfo` 
utility may be useful for you.

To list all current windows type the next command:

```
% xwininfo -tree -root
```

You may also get the window id for a specific window. Just run
`xwininfo` without arguments and point to the window.

```
% xwininfo
```

To pass the window id for *x11mirror-client* you have to use the `-w`
option, for instance:

```
% ./x11mirror-client -w 0x1400006
```

By default `x11mirror-client` uses the root window (whole screen).


### Case 1: save the current screenshot to a XWD file

In this case the program stores the screenshot of the window or the
whole display to one file. By default the outfile have the XWD format.
Use the `-o` option to specify a different location of the output file:

```
% ./x11mirror-client -o /path/to/x11mirror.xwd
```

The default output filepath is `/tmp/x11mirror.xwd`.


### Case 2: save the current screenshot to a gzip file

Same as above but with a compression of the outfile to gzip format.
The example below have an additional the `-z` option.

```
% ./x11mirror-client -o /path/to/x11mirror.xwd.gz -z
```

You may also specify a compression level via `-Z` option. For instance,
use the fastest compression:

```
% ./x11mirror-client -o /path/to/x11mirror.xwd.gz -z -Z 1
```

The default compression level is `7`. Note that if you not specify the output
filepath, then `x11mirror.xwd` will use `/tmp/x11mirror.xwd` without appended
`.gz` extension. This would not break `convert` utility to recognize the
correct file format.


### Case 3: send the current screenshot to a remote server via HTTP

You may combine the options above to get an acceptable result.
The *x11mirror-client* do a simple POST request to the remote server
using `multipart/form-data` format. To do so you have to specify
the `-U` option:

```
% ./x11mirror-client -U
```

To change the remote server URL you may use `-u` option, as show below:

```
% ./x11mirror-client -U -u http://example.net/
```

The format of the request which will be sent to the server is simple:

```
form-data;name="file";filename="x11mirror.xwd.gz"
[data bytes]
```

You may look at the [upload.c](/upload.c) file for details.


### Delay option

The delay option `-D` was introduced to save CPU resources of the computer.
The value should be specified as milliseconds. For instance, to get
screenshots each 2 seconds you have to run *x11mirror-client* in such way:

```
% ./x11mirror-client -D 2000
```

The default delay value is the 1 second (e.g. `-D 1000`).

If you would like to send screenshots without a delay you may re-build
the program with passing `ENABLE_DELAY=NO` to `make`.
See also [Makefile](/Makefile) for details.


## Why?

Just for fun.

To be more precise, when I had read the X11 documentation I have wanted to 
implement something unusual. Also such exercises helps remember what you 
had read about.


## How does it work?


### Introduction

The X11 or better to say the `xorg-server` is a complicated as hell thing.
It is trivial to retrieve a screenshot of the whole display. 
You may look at [xwd][3] utility to understand this yourself.

The things becomes complicated when you want to take a screenshot
of a minimized window. Or more common task is take a screenshot of
the window which is covered by other windows. That's why this 
`x11mirror` project was born.


### X11 concepts

Well, it is hard to explain all details in this README. I would say
that the most important part of X11 which is related to this project
is next: the X11 uses only one buffer to display content of windows.

That's right, there is no individual buffers for each window. When
the content of the window has draw by `xorg-server`, the last one
calculates the size of the window, its position, its intersections
with other windows (if they present) and then you see "a picture"
of the window.

In the case of intersections with other windows, the `xorg-server`
will not try to draw things that are covered by other windows.
You may imagine that `xorg-server` is operates under some
sort of a patchwork, where each part is either belongs to
the target window or does not.

That's why it is not possible to just get a screenshot
of the whole window if the last one is covered by other windows.

See links to docs below in the **References** section for details.


### X11 extentions

As was described above to take the screenshot of the whole window is
not possible by [xwd][3] if such a window is covered by other ones.

The *x11mirror-client* uses next X11 extentions:

* [COMPOSITE][4] - compose multiple windows into a one image
* [RENDER][5] - provides modern interface to draw pictures
* [DAMAGE][6] - monitor changes of windows


#### COMPOSITE

The role of COMPOSITE extention is to compose the output
images of the windows. It is also provides support of images
with alpha channel.

So, basically, if you want to draw images (content) of windows 
on the display in some kind of predefined order you would like to see, 
then you have to use this extention. For instance, if you wish 
to add true transparency to windows you have to use this extention 
too. There is no other way to get true transparency under 
X11 at the moment.

In `x11mirror-client` the COMPOSITE plays a role which
does the major job: it allows to get a content of whole window
even if it is covered by other ones.


#### RENDER

There is one issue which I did not mentioned before. Under
X11 to get an image (picture, content, you name it) of the
window you have care that Visual Format of the window
and the output image are the same. I will not provide
details about that here, you may look at them in **References**
section below.

So, the first thing is that RENDER extention will take care for
us about Visual Format of the window and the output image too.

Secondly, this extention provides an interface to for modern 
Picture format. The `x11mirror-client` uses it as an output buffer.


#### DAMAGE

This extention plays a helpful role. The X11 uses the event-driven model,
so instead of using polling method, the `x11mirror-client` asks 
the `xorg-server` send to it events when something is changed inside 
of the window. For instance, if a user is typing something in a browser 
(and new letters are drawn), the DAMAGE extention will tell about this.


#### Conclusions

In the end the overall logic of the program may put in this algorithm:

Input:

* Using common call XCreatePixmap() we create the Pixmap of the window.
* Using RENDER we create the intermediate Picture structure for the Pixmap of the window
  above.

Output:

* Using RENDER we create the output Picture structure.

Copying input to output:

* DAMAGE gets events when the content of the window is changed.
* When this is happen we copy the content of the window using RENDER to
  output Picture structure.
* DAMAGE is also looking at the content of output Picture. And when the
  copying is finished it sends another event about such changes.
* On event when copying is finished we store an image into XWD file.


One thing I forgot to tell. The X11 client program by design don't
have the content of the window in its own memory. To get the real
bytes of the image you have to use standard call to `XGetImage()`.
The function asks the `xorg-server` to transfer real bytes
to X11 client program. And later, the last one may do anything 
with them. In our case, save bytes to a file.

This function is cranky, it will not work properly if you pass
incorrect Visual Format to it. That's why RENDER is so helpful
to us, as I mentioned before.

That's all.




## See also

1. You may found the server to test the program here: [x11mirror-server][1]

2. [Keylogger][2] for X11


## Credits

Parts of the program contains a code provided by the [xwd][3] utility
wrote by The Open Group & Hewlett-Packard Co.


## References

The documentation for X11 is scattered over the Internet. Especially
in case of X11 extentions: the README included in these projects
provides a little details about how they work. So, I will try
to give you full information which may be interested in X11 development:


1. [The X New Developer’s Guide][7] - I would say that it provides an introduction to X11 world.

2. [Xlib - C Language X Interface][16] - xlib documentation, the main
document of X11 developer.

3. [XCB API][17] - if you'll plan to create X11 applications from a scratch, the xcb
provides a modern interface to X11 than xlib. You have to look on it.

4. [fredrik's X11 Composite Tutorial][8] - a good tutorial for practice
purposes.

5. [A New Rendering Model for X][11] by Keith Packard - a good article which explains why and
how RENDER extention was created.

6. [All articles by Keith Packard][12] - you have to read some of them too.
A lot of details were told on Usenix Conferences and probably you will not
find a place in Internet which explains X11 extentions in a such simple way.

7. [DAMAGE][9] - a document describes its API.

8. [RENDER][10] - a document describes its API.

9. `man xcomposite` - yep, manpage that describes its API.

10. [Everything curl][13] (pdf) - a wonderful book about CURL, also available
[online][14].

11. [zlib Manual][15] - explains itself.


[1]: https://github.com/gh0stwizard/x11mirror-server
[2]: https://github.com/xdevelnet/x11k
[3]: https://cgit.freedesktop.org/xorg/app/xwd/
[4]: https://cgit.freedesktop.org/xorg/lib/libXcomposite/
[5]: https://cgit.freedesktop.org/xorg/lib/libXrender/
[6]: http://cgit.freedesktop.org/xorg/proto/damageproto/
[7]: https://www.x.org/wiki/guide/
[8]: http://www.talisman.org/~erlkonig/misc/x11-composite-tutorial/
[9]: https://cgit.freedesktop.org/xorg/proto/damageproto/tree/damageproto.txt
[10]: https://cgit.freedesktop.org/xorg/lib/libXrender/tree/doc/libXrender.txt
[11]: https://keithp.com/~keithp/talks/usenix2000/render.html
[12]: https://keithp.com/~keithp/talks/
[13]: https://www.gitbook.com/download/pdf/book/bagder/everything-curl
[14]: https://ec.haxx.se/
[15]: http://www.zlib.net/manual.html
[16]: https://www.x.org/releases/current/doc/libX11/libX11/libX11.html
[17]: https://xcb.freedesktop.org/XcbApi/


Changes
0.2.0   Feb  1, 2021
    - Improved build process
    - Added the option to control uploading process (-U)
    - Updated readme

0.1     Mar 24, 2017
    - Initial release


main.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <limits.h>
#include <getopt.h>
#include <stdarg.h>
#include <sys/select.h>
#include <sys/time.h>
#include <sys/types.h>
#ifdef USE_DELAY
    #include <time.h>
    #include <math.h>
#endif
#include <unistd.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xatom.h>
#include <X11/extensions/Xcomposite.h>
#include <X11/extensions/Xrender.h>
#include <X11/extensions/Xdamage.h>
#include <X11/extensions/shape.h>
#include "common.h"
#include "window_dump.h"
#ifdef HAVE_ZLIB
    #include "compression.h"
#endif
#ifdef HAVE_CURL
    #include "upload.h"
#endif

/* is there compositor manager */
static Bool test_cm (Display *d);
/* load necessary x11's extensions */
static void load_x11_extensions (Display *);
static void debug_window (Display *d, Window w, XWindowAttributes *wa);
/* select events for specified target window */
static void xselect_input (Display *d, Window w);
static inline void
save_file (Display *d, int screen, Window w, Pixmap p);
/* by default X11 throws error and exit */
static int
xerror_handler (Display *dpy, XErrorEvent *ev);

/* X11 extensions offsets */
static int composite_opcode, composite_event, composite_error;
static int render_event, render_error;
static int xfixes_event, xfixes_error;
static int damage_event, damage_error;
/*static int xshape_event, xshape_error;*/

static char *out_file_name;
static FILE *out_file;

#define _STR(x) #x
#define STR(x) _STR(x)

#ifdef USE_DELAY
#ifndef _DELAY_NSEC
/* very low numbers may lead to lags */
#define _DELAY_NSEC 0 * 1000000
#endif
#ifndef _DELAY_SEC
#define _DELAY_SEC 1
#endif
#endif

#ifndef _OUTPUT_FILE
#define _OUTPUT_FILE /tmp/x11mirror.xwd
#endif

#ifdef HAVE_ZLIB
#ifndef _ZLIB_DEFAULT_LEVEL
#define _ZLIB_DEFAULT_LEVEL 7
#endif
static Bool use_zlib = False;
static unsigned zlevel = _ZLIB_DEFAULT_LEVEL;
#endif

#ifdef HAVE_CURL
#ifndef _UPLOAD_URL
#define _UPLOAD_URL "http://localhost:8888"
#endif
static Bool enable_upload = False;
#endif


static void
print_usage (const char *prog)
{
    fprintf (stderr, "Usage: %s [-w window] [OPTIONS]\n", prog);
    fprintf (stderr, "Options:\n");
#define desc(o,d) fprintf (stderr, "  %-16s  %s\n", o, d);
    desc ("--help, -h", "print this help");
    desc ("--version, -v", "print the program version");
    desc ("-d display", "connection string to X11");
    desc ("-o output", "output filename, default: " STR(_OUTPUT_FILE));
    desc ("-w window", "target window id, default: root");
    desc ("-S", "enable X11 synchronization, default: disabled");
#ifdef HAVE_CURL
    desc ("-u URL", "an URL to send data");
    desc ("-U", "enable uploading, default: disabled");
#endif
#ifdef HAVE_ZLIB
    desc ("-z", "enable gzip (zlib) compression, default: disabled");
    desc ("-Z", "zlib compression level (1-9), default: " STR(_ZLIB_DEFAULT_LEVEL));
#endif
#ifdef USE_DELAY
    desc ("-D", "delay between making screenshots in milliseconds");
#endif
#undef desc
}


static void
print_version ()
{
    const char futures[] = {
#ifdef HAVE_CURL
        " +curl"
#endif
#ifdef HAVE_ZLIB
        " +zlib"
#endif
#ifdef _DEBUG
        " +debug"
#endif
#ifdef PRINT_ERRORS
        " +errors"
#endif
#ifndef USE_DELAY
        " -delay"
#endif
    };

    printf ("v%s%s\n", STR(APP_VERSION), futures);
}


extern int
main (int argc, char *argv[])
{
    int         opt;                /* options of  getopt */
    Window      w = 0;              /* target window */
    int         w_screen = -1;      /* number of screen of target window */
    Display     *dpy;
    char        *display = NULL;    /* display string from args */
    XEvent      ev;
    XWindowAttributes wa;
    Pixmap      w_pixmap = 0;
    XRenderPictFormat *format;
    Picture     w_picture, pixmap_picture;
    XRenderPictureAttributes w_pa = { .subwindow_mode = IncludeInferiors };
    Damage      w_damage, pixmap_damage;
    Bool        synchronize = False;
    unsigned    my_events;
#ifdef USE_DELAY
    time_t      delay_sec = _DELAY_SEC;
    long        delay_nsec = _DELAY_NSEC;
#endif
#ifdef HAVE_CURL
    char        *url = NULL;
#endif

    static struct option opts[] = {
        { "help",       no_argument,        0, 'h' },
        { "version",    no_argument,        0, 'v' },
        { "display",    required_argument,  0, 'd' },
        { "output",     required_argument,  0, 'o' },
        { "window",     required_argument,  0, 'w' },
        { "sync-x11",   no_argument,        0, 'S' },
#ifdef HAVE_CURL
        { "url",        required_argument,  0, 'u' },
        { "upload",     no_argument,        0, 'U' },
#endif
#ifdef HAVE_ZLIB
        { "compress",   no_argument,        0, 'z' },
        { "z-level",    required_argument,  0, 'Z' },
#endif
#ifdef USE_DELAY
        { "delay",      required_argument,  0, 'D' },
#endif
        { 0, 0, 0, 0 }
    };

    while (1) {
        int index = 0;
        opt = getopt_long (argc, argv, "hvd:o:w:Su:UzZ:D:", opts, &index);
        if (opt == -1) break;
        switch (opt) {
        case 'd':
            display = optarg;
            break;
#ifdef HAVE_ZLIB
        case 'z':
            use_zlib = True;
            break;
        case 'Z':
            sscanf (optarg, "%u", &zlevel);
            if (zlevel <= 0 || zlevel >= 10)
                die ("Invalid zlib compression level: %s.", optarg);
            break;
#endif
        case 'o':
            out_file_name = optarg;
            break;
#ifdef HAVE_CURL
        case 'u':
            url = optarg;
            break;
        case 'U':
            enable_upload = True;
            break;
#endif
        case 'w':
            sscanf (optarg, "0x%lx", &w);
            if (w == 0)
                sscanf (optarg, "%lu", &w);
            if (w == 0)
                die ("Invalid window id: %s.", optarg);
            break;
#ifdef USE_DELAY
        case 'D': {
            long delay = 0;
            sscanf (optarg, "%li", &delay);
            if (delay < 0)
                die ("Delay must be greater than zero.");
            else if (delay >= 1000) {
                delay_sec = lround ((delay - (delay % 1000)) / 1000);
                delay_nsec = (delay % 1000) * 1000000;
            }
            else {
                delay_sec = 0;
                delay_nsec = delay * 1000000;
            }
        }   break;
#endif
        case 'S':
            synchronize = True;
            break;
        case 'h':
            print_usage (argv[0]);
            exit (EXIT_SUCCESS);
        case 'v':
            print_version ();
            exit (EXIT_SUCCESS);
        default: /* '?' */
            print_usage (argv[0]);
            exit (EXIT_FAILURE);
        }
    }

    if (out_file_name == NULL) {
        out_file_name = STR(_OUTPUT_FILE);
    }
    else {
        int len = strlen (out_file_name);

        if (len <= 0 && len >= PATH_MAX)
            die ("invalid length of output filename.");
    }

    debug ("output file: %s\n", out_file_name);
#ifdef HAVE_ZLIB
    debug ("output file format: %s\n", (use_zlib) ? "zlib" : "xwd");
#endif


#ifdef HAVE_CURL
    /* initialize curl first */
    if (url == NULL)
        url = _UPLOAD_URL;

    if (enable_upload) {
        debug ("uploading to %s\n", url);
        init_uploader (url);
    }
    /* open file in read-write mode */
    out_file = fopen (out_file_name, "w+b");
#else
    /* otherwise we need only write permissions */
    out_file = fopen (out_file_name, "wb");
#endif

    if (out_file == NULL) {
        perror ("fopen output filename");
        exit (EXIT_FAILURE);
    }

#ifdef USE_DELAY
    debug ("delay between screenshots: %lu.%lu second(s)\n",
            delay_sec, lround (delay_nsec / 1000000));
#endif

    dpy = XOpenDisplay (display);
    if (!dpy)
        die ("Could not open the display.");

    if (synchronize) {
        XSynchronize (dpy, synchronize);
        debug ("WARNING: X11 synchronization enabled\n");
    }

    XSetErrorHandler (xerror_handler);

    load_x11_extensions (dpy);
    const int damageEventType = damage_event + XDamageNotify;

    if (w == 0) {
        w = RootWindow (dpy, DefaultScreen (dpy));
        debug ("using root window as target one.\n");
    }

    if (test_cm (dpy))
        debug ("compositor manager was not found.\n");
    else
        debug ("compositor manager has been found.\n");

    for (int i = 0; i < ScreenCount (dpy); i++) {
        XCompositeRedirectSubwindows (dpy, RootWindow (dpy, i),
            CompositeRedirectAutomatic);
    }

    debug ("retrieving attributes for 0x%lx\n", w);
    XGetWindowAttributes (dpy, w, &wa);
    debug_window (dpy, w, &wa);

    for (int i = 0; i < ScreenCount (dpy); i++) {
        if (ScreenOfDisplay (dpy, i) == wa.screen) {
            w_screen = i;
            debug ("default screen = %d, target window screen = %d\n",
                DefaultScreen (dpy), i);
            break;
        }
    }

    if (w_screen < 0)
        die ("failed to find out a screen number of the window.");

    /* select events for the window */
    xselect_input (dpy, w);

    format = XRenderFindVisualFormat (dpy, wa.visual);
    w_picture = XRenderCreatePicture (dpy, w, format, CPSubwindowMode, &w_pa);
    w_pixmap = XCreatePixmap (dpy, w, wa.width, wa.height, wa.depth);
    pixmap_picture = XRenderCreatePicture (dpy, w_pixmap, format, 0, NULL);

    w_damage = XDamageCreate (dpy, w, XDamageReportRawRectangles);
    pixmap_damage = XDamageCreate (dpy, w_pixmap, XDamageReportRawRectangles);

    XFlush (dpy);

#define COMPOSITE() XRenderComposite (dpy, \
    PictOpSrc, w_picture, None, pixmap_picture, \
    0, 0, 0, 0, \
    0, 0, wa.width, wa.height)

    debug ("waiting for events ...\n");

    while (1) {
        my_events = 0;

        do {
            XNextEvent (dpy, &ev);

            debug ("event for 0x%08lx type: %4d", ev.xany.window, ev.type);

            if (ev.type == damageEventType) {
                Damage d = ((XDamageNotifyEvent *)&ev)->damage;

                if (d == w_damage) {
                    debug (" damage event for the window\n");
                    my_events |= 1 << 0;
                }
                else if (d == pixmap_damage) {
                    debug (" damage event for the pixmap\n");
                    my_events |= 1 << 1;
                }

                continue;
            }
            else
                debug ("\n");

            switch (ev.type) {
            case ConfigureNotify:
            case MapNotify: {
                if (ev.xconfigure.window == w) {
                    debug ("updating window attributes.\n");
                    XGetWindowAttributes (dpy, w, &wa);
                    format = XRenderFindVisualFormat (dpy, wa.visual);
                    XFreePixmap (dpy, w_pixmap);
                    XRenderFreePicture (dpy, pixmap_picture);
                    XRenderFreePicture (dpy, w_picture);
                    XDamageDestroy (dpy, pixmap_damage);
                    w_picture = XRenderCreatePicture
                        (dpy, w, format, CPSubwindowMode, &w_pa);
                    w_pixmap = XCreatePixmap
                        (dpy, w, wa.width, wa.height, wa.depth);
                    pixmap_picture = XRenderCreatePicture
                        (dpy, w_pixmap, format, 0, NULL);
                    pixmap_damage = XDamageCreate
                        (dpy, w_pixmap, XDamageReportRawRectangles);
                }
            }   break;
            case DestroyNotify:
                if (ev.xunmap.window == w)
                    goto done;
                break;
            } // switch (ev.type) {
        } while (QLength (dpy));


        if (my_events == 0) {
            int fd = ConnectionNumber (dpy);
            fd_set fdset;
            FD_ZERO (&fdset);
            FD_SET (fd, &fdset);
            struct timeval wtime = { 0, 500000 };
            select (fd + 1, &fdset, NULL, NULL, &wtime);
        }
        else if (my_events & (1 << 1)) {
            save_file (dpy, w_screen, w, w_pixmap);
#ifdef HAVE_CURL
            if (enable_upload)
                upload_file (out_file);
#endif

#ifdef USE_DELAY
            struct timespec wtime = {
                .tv_sec = delay_sec,
                .tv_nsec = delay_nsec
            };
            nanosleep (&wtime, NULL);
#endif
        }
        else if (my_events & (1 << 0)) {
            COMPOSITE ();
        }
    } /* while (1) */

#undef COMPOSITE

done:
    if (w_pixmap)
        XFreePixmap (dpy, w_pixmap);
    XDamageDestroy (dpy, w_damage);
    XDamageDestroy (dpy, pixmap_damage);
    XRenderFreePicture (dpy, pixmap_picture);
    XRenderFreePicture (dpy, w_picture);
    XSync (dpy, False);
    XCloseDisplay (dpy);
#ifdef HAVE_CURL
    free_uploader ();
#endif
    fclose (out_file);

    return 0;
}


static void
load_x11_extensions (Display *dpy)
{
    int major_version, minor_version;

    if (!XQueryExtension (dpy, COMPOSITE_NAME, 
        &composite_opcode, &composite_event, &composite_error))
        die ("COMPOSITE extension is not installed.");
    XCompositeQueryVersion (dpy, &major_version, &minor_version);
    if (major_version == 0 && minor_version < 4)
        die ("libXcomposite 0.4 or better is required.");
    debug ("using COMPOSITE %d.%d\n", major_version, minor_version);

    if (!XRenderQueryExtension (dpy, &render_event, &render_error))
        die ("RENDER extension is not installed.");
    XRenderQueryVersion (dpy, &major_version, &minor_version);
    debug ("using RENDER %d.%d\n", major_version, minor_version);

    if (!XFixesQueryExtension (dpy, &xfixes_event, &xfixes_error))
        die ("FIXES extension is not installed.");
    XFixesQueryVersion (dpy, &major_version, &minor_version);
    debug ("using FIXES %d.%d\n", major_version, minor_version);

    if (!XDamageQueryExtension (dpy, &damage_event, &damage_error))
        die ("DAMAGE extension is not installed.");
    XDamageQueryVersion (dpy, &major_version, &minor_version);
    debug ("using DAMAGE %d.%d\n", major_version, minor_version);
}


static void
xselect_input (Display *d, Window w)
{
    Window      root, parent, *children;
    unsigned int nchildren;
    long ev_mask = StructureNotifyMask | SubstructureNotifyMask | ExposureMask;

    XGrabServer (d);
    XSelectInput (d, w, ev_mask);
    XQueryTree (d, w, &root, &parent, &children, &nchildren);

    for (unsigned int i = 0; i < nchildren; i++) {
        XSelectInput (d, children[i], ev_mask);
    }

    XFree (children);
    XUngrabServer (d);
}


static void
debug_window (Display *d, Window w, XWindowAttributes *wa)
{
#define BACKSTORE(i) ( (i == NotUseful) \
    ? ("Not Useful") \
    : (i == WhenMapped) ? ("When Mapped") : ("Always") )

    (void)wa;
    char *window_name = NULL;

/* TODO: _NET_WM_NAME */
    XFetchName (d, w, &window_name);

    if (window_name != NULL) {
        debug ("  name: %s\n", window_name);
        XFree (window_name);
    }

    debug ("  x: %d y: %d\n", wa->x, wa->y);
    debug ("  width: %d height: %d\n", wa->width, wa->height);
    debug ("  root: 0x%lx\n", wa->root);
    debug ("  backing store: %s\n", BACKSTORE(wa->backing_store));
#undef BACKSTORE
}


/*
 * test_cm: returns True if there is no any Compositor Manager.
 */
static Bool
test_cm (Display *dpy)
{
    static char name[] = "_NET_WM_CM_Sxx";

    snprintf (name, sizeof (name), "_NET_WM_CM_S%d", DefaultScreen (dpy));

    if (None == XGetSelectionOwner (dpy, XInternAtom (dpy, name, False)))
        return True;
    else
        return False;
}


static inline void
save_file (Display *d, int screen, Window w, Pixmap p)
{
    static mirrorDump *dump;

    if ((dump = Pixmap_Dump (d, screen, w, p)) != NULL) {
        rewind (out_file);
#ifdef HAVE_ZLIB
        if (use_zlib)
            save_gzip_file (dump, out_file, zlevel);
        else
#endif
            Save_Dump (dump, out_file);
    }

    Free_Dump (dump);
    fflush (out_file);
}


static int
xerror_handler (Display *dpy, XErrorEvent *ev)
{
#if defined(_NO_ERRORS)
    (void) dpy;
    (void) ev;
#else
    int         o;
    const char  *name = NULL;
    static char buffer[256];

    o = ev->error_code - xfixes_error;
    switch (o) {
    case BadRegion: name = "BadRegion"; break;
    default: break;
    }

    o = ev->error_code - damage_error;
    switch (o) {
    case BadDamage: name = "BadDamage"; break;
    default: break;
    }

    o = ev->error_code - render_error;
    switch (o) {
    case BadPictFormat: name ="BadPictFormat";  break;
    case BadPicture:    name ="BadPicture";     break;
    case BadPictOp:     name ="BadPictOp";      break;
    case BadGlyphSet:   name ="BadGlyphSet";    break;
    case BadGlyph:      name ="BadGlyph";       break;
    default: break;
    }

    if (name == NULL) {
        buffer[0] = '\0';
        XGetErrorText (dpy, ev->error_code, buffer, sizeof (buffer));
        name = buffer;
    }

    fprintf (stderr, "error %d: %s request %d minor %d serial %lu\n",
        ev->error_code, (strlen (name) > 0) ? name : "unknown",
        ev->request_code, ev->minor_code, ev->serial);
#endif

    return 0;
}
